### PR TITLE
Fixing issue #5376 by adding a check to parse out Zone info

### PR DESCRIPTION
**Pull request by SamrathPalSingh**
This fixes issue #5376 by simply removing any Zone info in the IP address before net.ParseIP(). This also adds a test in the acl-test  file to test the new changes.

It also blocks any queries which fail in the parsing step.

### CODE REVIEW

**Comment by SamrathPalSingh** [comment]
@chrisohaver I'm not so sure about the second change, where we block any requests which fail the parsing step. 
```
	if len(ip) == 0 {
		return actionBlock
	}
```
Should I remove/edit it ?

File: plugin/acl/acl.go [hunk 1]
```diff
@@ -96,7 +98,13 @@ RulesCheckLoop:
 func matchWithPolicies(policies []policy, w dns.ResponseWriter, r *dns.Msg) action {
 	state := request.Request{W: w, Req: r}
 
-	ip := net.ParseIP(state.IP())
+	ip := net.ParseIP(strings.Split(state.IP(), "%")[0])
+
+	// if the parsing did not return a proper response then we simply return 'actionBlock' to
+	// block the query
+	if len(ip) == 0 {
```
Comment (chrisohaver) [inline plugin/acl/acl.go]: Probably better to compare the result to nil to check if parsing failed. i.e. `if ip == nil {`

File: plugin/acl/acl.go [hunk 2]
```diff
@@ -96,7 +98,13 @@ RulesCheckLoop:
 func matchWithPolicies(policies []policy, w dns.ResponseWriter, r *dns.Msg) action {
 	state := request.Request{W: w, Req: r}
 
-	ip := net.ParseIP(state.IP())
+	ip := net.ParseIP(strings.Split(state.IP(), "%")[0])
```
Comment (chrisohaver) [inline plugin/acl/acl.go]: It's probably slightly less expensive to use `strings.IndexByte` in place of `Split` (assuming Go compiler doesn't optimize based on the literal `[0]`)... e.g. something like the following:

```go
if idx := strings.IndexByte(state.IP(), '%'); idx >= 0 {
    ip = net.ParseIP(state.IP()[:idx])
}

File: plugin/acl/acl.go [hunk 3]
```diff
@@ -96,7 +98,13 @@ RulesCheckLoop:
...
+
+	// if the parsing did not return a proper response then we simply return 'actionBlock' to
+	// block the query
+	if len(ip) == 0 {
+		return actionBlock
```
Comment (chrisohaver) [inline plugin/acl/acl.go:114]: If we can't parse the trimmed IP string, we should log an error.
I agree that blocking is probably the best action here.

**Review by chrisohaver — CHANGES_REQUESTED** [review]
Thanks!

File: plugin/acl/acl.go (hunk 3 above)
Comment (SamrathPalSingh) [reply plugin/acl/acl.go:114]: @chrisohaver - added a log statement which mentions the failing IP

File: plugin/acl/acl.go [hunk 4]
```diff
@@ -96,7 +99,19 @@ RulesCheckLoop:
 func matchWithPolicies(policies []policy, w dns.ResponseWriter, r *dns.Msg) action {
 	state := request.Request{W: w, Req: r}
 
-	ip := net.ParseIP(state.IP())
+	ip := []byte(nil)
```
Comment (chrisohaver) [inline plugin/acl/acl.go]:
```suggestion
	var ip net.IP
```

File: plugin/acl/acl.go [hunk 5]
```diff
@@ -96,7 +99,19 @@ RulesCheckLoop:
...
+	if idx := strings.IndexByte(state.IP(), '%'); idx >= 0 {
+		ip = net.ParseIP(state.IP()[:idx])
+	} else {
+		ip = net.ParseIP(state.IP())
+	}
+
+	// if the parsing did not return a proper response then we simply return 'actionBlock' to
+	// block the query
+	if ip == nil {
+		log.Errorf("ParseIP failed for ip address: %v", state.IP())
```
Comment (chrisohaver) [inline plugin/acl/acl.go]: Please use the plugin logger  https://github.com/coredns/coredns/blob/master/plugin/pkg/log/plugin.go. It logs the plugin name along with the message.

example: https://github.com/coredns/coredns/blob/master/plugin/cache/setup.go#L17

A better error message would be "Blocking request. Unable to parse source address: ".

**Review by chrisohaver — CHANGES_REQUESTED** [review]

File: plugin/acl/acl.go (hunk 4 above)
Comment (SamrathPalSingh) [reply plugin/acl/acl.go]: Just for my knowledge, is there an actual difference between using the var keyword and the short declaration operator in this particular case ?

Changing []byte to net.IP, I realize is safer.

**Review by chrisohaver — COMMENTED** [review]
Need to fix the import ordering/grouping, but looks good otherwise.  (this is also just a style thing, but it is enforced in the tests)

File: plugin/acl/acl.go (hunk 4 above)
Comment (chrisohaver) [reply plugin/acl/acl.go]: no difference afaik.

IMO, if declaring a new empty variable, var is preferred style. If also assigning a value at the same time, then := is fine.

**Comment by SamrathPalSingh** [comment]
@chrisohaver fixed the import ordering. All check have passed.

**Review by chrisohaver — APPROVED** [review]
Lgtm

**Review by yongtang — APPROVED** [review]

**Comment by ihac** [comment]
LGTM

**Pull request merged**